File: java/src/main/java/com/danifoldi/dml/type/DmlKey.java
package com.danifoldi.dml.type;

import com.danifoldi.dml.DmlSerializer;

import java.util.Objects;

public class DmlKey extends DmlCommentableValue implements Comparable<DmlKey> {
    private String value;

    public DmlKey(String value) {
        this.value = value;
    }

    public String value() {
        return this.value;
    }

    public void value(String value) {
        this.value = value;
    }

    @Override
    public String serialize(int indent) {
        return DmlSerializer.serializeComment(comment(), indent) + DmlSerializer.indent(value, indent);
    }

    @Override
    public String toString() {
        return "DmlKey{" + value + "}";
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        DmlKey dmlKey = (DmlKey) o;
        return Objects.equals(value, dmlKey.value);
    }

    @Override
    public int hashCode() {
        return Objects.hash(value);
    }

    @Override
    public int compareTo(DmlKey o) {
        return value.compareTo(o.value);
    }
}

File: javascript/src/types/DmlKey.js
import DmlCommentableValue from './DmlCommentableValue.js'
import DmlSerializer from '../DmlSerializer.js'

class DmlKey extends DmlCommentableValue {
  #value

  constructor(value) {
    super()
    this.#value = value
  }

  value(value) {
    if (typeof value === 'undefined') {
      return this.#value
    }

    this.#value = value
  }

  serialize(indent) {
    return DmlSerializer.serializeComment(this.comment(), indent) + DmlSerializer.indent(this.#value, indent)
  }
}

export default DmlKey
